2FA Login

cconard96 opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

- [x] Base feature
- [ ] Backup codes
- [ ] Administration (policy control)

orthagh reviewed (changes requested):
Some issues i found during 1st tests:

- when enabling "enforcing 2FA" and directly logout/login, i has been asked for a code but i didn't setup my application. There should a scenario where a user doesn't setup should have the registering form and not the seondary login process.
- maybe linked, the initial redirection to `?mfa_setup=1 ` redirect to base root of my localhost domain . I am in a subfolder (http://localhost/glpi/.../) and i need to rewrite manually the url.
- some UX enhancements to do:
   - the registering page could be prettier and guide more the user, some example
      - https://www.beyondtrust.com/docs/remote-support/resources/images/2fa/2fa-replace-21-3_thumb_0_0.png
      - https://dribbble.com/shots/15641874--2FA
      - https://miro.medium.com/v2/format:webp/1*En1MJwdo-U8Rl-voliThfA.png
   - same for recovery code, enhance UI please
   - pasting (with Ctrl+v) a 6 digits code in the intended input should fill all the inputs, not only the first one
   - [use a backup code] button should be on the same line as others ([cancel])
   - Auto-submiting the form after filling the 6-digits code is good, but depsite it's not usefull, a [Submit] button should be displayed to have a comprehensive design (and provide a fallback if something break)
- security tab in Entity miss an icon
- i didn't see this tab in Profile/group (this has been specificaly asked by the customer)
- not sure where is the bug, but the manual registering (with copy/pasting the key in my Authy app) doesn't to be functional. The provided 6-digits code is not accepted by glpi this way. By using QR-code, it's fine.

orthagh commented:
Another things:

- a cli command to manage a user 2FA could be usefull. Imagine you lost the 2fa for the super-admin account, you're fucked (need to alter things in DB)
- the searchoption for 2FA status in user list may be searchable

cconard96 commented:
> - the searchoption for 2FA status in user list may be searchable

No support for searching virtual search options.

flegastelois commented on src/Security/TOTPManager.php:
```diff
@@ -0,0 +1,379 @@
+
+    public const CODE_VALIDITY_SECONDS = 30;
+
+    public const CODE_ALGORITHM = 'sha1';
```
SHA1 should not be avoided? since 2015, it is an algorithm that is considered insecure.

cconard96 replied:
Yes it should be. However, Google Authenticator doesn't work when I tried SHA-256. I spent a lot of time trying to figure out why it wasn't working with it.

cconard96 replied:
An important note is that the RFC for TOTP (rfc6238) is based on HOTP (rfc4226) which specifically mentions SHA-1. By default, authenticator apps seem to use SHA-1 by default for compatibility reasons. When you use a QR code instead of a manually entered secret, it is actually a `otpauth` URI which can contain an `algorithm` parameter to specify the use of a different algorithm. For some reason, the Google Authenticator still ignores the parameter in the URI (I verified it was set correctly). I don't know why we should expect them to support a parameter of a URI they formalized themselves...
https://github.com/google/google-authenticator/wiki/Key-Uri-Format#algorithm

As for the real world implications, there are people who know way more than me on the subject. However, from what I understand, TOTP uses HMAC (HMAC-SHA-1 in this case) which does reduce the risks of using SHA-1 as the secret key is assumed to be unknown by anyone except the application and client.

cconard96 commented:
>     There should a scenario where a user doesn't setup should have the registering form and not the seondary login process.

That is already the case. I broke the enforced setup twig template with a rouge parenthesis. In some odd cases (bugs), the 2FA code request screen can show instead of the enforced setup screen (asks users to set up 2FA or skip if within the grace period), but that is not a normal behavior and the user can cancel out of it if it ever ends up in that state.

>     * not sure where is the bug, but the manual registering (with copy/pasting the key in my Authy app) doesn't seems to be functional. The provided 6-digits code is not accepted by glpi this way. By using QR-code, it's fine.

I can't find any proper documentation on how the Authy app works. I'm guessing that the default algorithm is different than SHA-1. When using the QR code, you can specify the algorithm used to generate the TOTP codes. Unfortunately, Google Authenticator is a very popular authenticator and they formalized the URI spec used in QR code setups, but ignore their own algorithm parameter and only support SHA-1,

cconard96 commented:
Some technical changes made:
1. Fixed redirects, missing icon, and missing enforcement configurations
2. Added logic to automatically determine the algorithm used by the authenticator in case it isn't sha1 like we expected. When setting the secret, all supported algorithms are checked (sha1 and sha256 currently) and the matching one is stored in the user's 2FA config for future use. This should improve compatibility when using the manual secret entry.

orthagh commented:
I confirm the authy registering and the 1st enroll are succesfull now :+1:

cconard96 commented:
UX changes made.

Regular/optional 2FA Setup:
![Selection_128](https://user-images.githubusercontent.com/17678637/217022594-4f594aa7-3070-4704-a532-6f2b46fa58bd.png)

Enforced setup during login (in grace period):
![Selection_131](https://user-images.githubusercontent.com/17678637/217022679-7e94230e-bb59-42e7-bab9-84b60266a34f.png)

Enforced setup during login (no/outside grace period):
![Selection_130](https://user-images.githubusercontent.com/17678637/217022660-a206b62d-717e-46a3-8e89-f8ec0c78785d.png)

2FA code request:
![Selection_132](https://user-images.githubusercontent.com/17678637/217022692-155df844-4235-4c15-9307-7c94f082ae55.png)

2FA backup code
![Selection_133](https://user-images.githubusercontent.com/17678637/217022715-4faa9a31-9094-4346-8a11-56e564c5485c.png)
 request:

2FA backup code generation:
![Selection_129](https://user-images.githubusercontent.com/17678637/217022620-fb8b6551-95f1-48e0-bf45-5363d0959c43.png)

orthagh commented:
Seems very good.
To be picky, i'll suggest:

- to avoid repetitions ("2FA setup", "2FA is required for your account")
- to add some vertical spacing around your numbered paragraphs (this is valid for any explanation paragraph, keep them "aerated")
- to avoid rounded borders on digit inputs not placed on the edges or **BETTER**:
   - use spacing between digits (keep rounded borders)
   - increase greatly font and input sizes (this is the primary control, we want it to be the attraction of the page)
   - ex:

![image](https://user-images.githubusercontent.com/418844/217026893-4e84b907-3c45-46e3-96eb-f98ed5f68b59.png)

cconard96 commented:
Done

orthagh reviewed (changes requested):
found some issues while testing correctly:

- copy button right of code (when registering) doesn't seem to work (both chrome and firefox) and a js error is present
  ```
  Uncaught (in promise) TypeError: Failed to execute 'write' on 'Clipboard': The object must have a callable @@iterator property.
  ```
- .card html issue:
    ![image](https://user-images.githubusercontent.com/418844/217035173-20b0bbcb-0173-400e-886a-6be8f04dd5fd.png)
- I didn't understand [reset 2FA] button purpose in preferences, nothing seems to change after clicking to it


UI/UX pov, nothing more to say, great job !

Tomorrow, I'll check/review Entity/Profile/Group enforcing, now i need to go

cconard96 commented:
>     * I didn't understand [reset 2FA] button purpose in preferences, nothing seems to change after clicking to it
It is supposed to display the setup form again without actually disabling 2FA first. It shows in place of the Disable 2FA button when 2FA is mandatory. It is bugged because the GET parameter that is set is not available when the tab content loads.

cedric-anne commented on src/Config.php:
```diff
@@ -3415,6 +3427,42 @@ public function showFormSecurity()
+            $days_left = (new \Glpi\Security\TOTPManager())->getGracePeriodDaysLeft();
+            if ($days_left >= 0) {
+                $end_ts = strtotime($CFG_GLPI['2fa_grace_date_start']) + ($CFG_GLPI['2fa_grace_days'] * DAY_TIMESTAMP);
+                echo '<br><span>' . __('2FA grace period started: ') . Html::convDateTime($CFG_GLPI['2fa_grace_date_start']) . '</span>';
```
Value position may change in the sentence depending on the locale.
```suggestion
                echo '<br><span>' . sprintf(__('2FA grace period started on %s'), Html::convDateTime($CFG_GLPI['2fa_grace_date_start'])) . '</span>';
```

cedric-anne commented on src/Console/Security/DisableTFACommand.php:
```diff
@@ -0,0 +1,46 @@
+        parent::configure();
+        $this->setName('security:disable_2fa');
+        $this->setDescription(__('Disable 2FA for a user'));
+        $this->addArgument('login', InputArgument::REQUIRED, 'Username');
```
```suggestion
        $this->addArgument('login', InputArgument::REQUIRED, __('Username'));
```

cedric-anne commented on src/Console/Security/DisableTFACommand.php:
```diff
@@ -0,0 +1,46 @@
+        $username = $input->getArgument('login');
+        $user = new \User();
+        if (!$user->getFromDBbyName($username)) {
+            $output->writeln("<error>" . __("User $username not found") . "</error>");
```
```suggestion
            $output->writeln("<error>" . sprintf(__("User %s not found"), $username) . "</error>");
```

cedric-anne commented on src/Console/Security/DisableTFACommand.php:
```diff
@@ -0,0 +1,46 @@
+            return 0;
+        }
+        $totp_manager->disable2FAForUser($user->getID());
+        $output->writeln("<info>" . __("2FA disabled for user $username") . "</info>");
```
```suggestion
        $output->writeln("<info>" . sprintf(__("2FA disabled for user %s"), $username) . "</info>");
```

cedric-anne commented on src/Profile.php:
```diff
@@ -2150,6 +2156,58 @@ public function showFormSetupHelpdesk($openform = true, $closeform = true)
+
+        $rand = mt_rand();
+
+        echo "<tr><th colspan='2'>" . __('2FA') . "</th></tr>";
```
Same remark.

cedric-anne commented on src/Security/TOTPManager.php:
```diff
@@ -0,0 +1,495 @@
+            }
+            if (!$enforced) {
+                // Check profile configuration
+                $profiles = \Profile_User::getUserProfiles($users_id);
+                if (!empty($profiles)) {
+                    $enforced = $DB->request([
+                        'SELECT' => ['2fa_enforced'],
+                        'FROM' => 'glpi_profiles',
+                        'WHERE' => [
+                            'id' => $profiles,
+                            '2fa_enforced' => 1
+                        ]
+                    ])->count() > 0;
+                }
```
This implies that a user that has a limited profile, that does not enforce 2FA, as its default profile and an admin profile, that enforces 2FA, and that is used really often, then the user will be forced to use 2FA on both profiles. Is it possible to activate the 2FA only on profile selection?

cedric-anne replied:
We discussed about it with @orthagh.

Requesting a OTP code on login as long as user has access to an entity that enforces 2FA or is part of a group/profile that enforces 2FA is the expected behaviour.

The only thing to add it a :information_source: block on corresponding configuration forms to clearly explains this.

cedric-anne commented on templates/pages/2fa/2fa_status.html.twig:
```diff
@@ -0,0 +1,133 @@
+      </h2>
+   </div>
+   <div id="backup_codes" class="d-none py-2" style="width: fit-content;">
+      <h3 class="mb-2">{{ __('Backup codes (This is the only time these will be shown)') }}</h3>
```
Maybe the `This is the only time these will be shown` message should be shown in a :warning: block.

cconard96 commented on src/Config.php:
```diff
@@ -3415,6 +3427,42 @@ public function showFormSecurity()
+        echo '<td>';
+        Dropdown::showNumber('2fa_grace_days', ['value' => $CFG_GLPI["2fa_grace_days"],
+            'min'   => 0,
+            'max'   => 365,
```
That's not really for me to say. It is up to the IT/Security team.

cconard96 commented on src/Auth.php:
```diff
@@ -942,14 +960,84 @@ public function login($login_name, $login_password, $noauto = false, $remember_m
+        if ($mfa_pre_auth || $this->validateLogin($login_name, $login_password, $noauto, $login_auth)) {
+            // Check MFA
+            $totp = new TOTPManager();
+            $web_access = !isAPI() && !isCommandLine();
```
I don't know the CalDAV feature at all, but basic authentication probably shouldn't be supported in the API at all and can be disabled with the pending OAuth server and High-Level API feature PRs.

cedric-anne replied:
We discussed about it with @orthagh .

CalDAV and API security of API/CalDAV endpoints will be enforced in dedicated PRs. We added corresponding tasks in the roadmap.

I pushed a commit that fixes CalDAV endpoint authentication.

cedric-anne reviewed (commented):
Looks OK to me. I would not have time to do extensive testing right now since I will be off for a week.

orthagh commented on src/Security/TOTPManager.php:
```diff
@@ -0,0 +1,495 @@
+            // Check entity configuration
+            $entities = PermissionManager::getInstance()->getAllEntities($users_id);
+            foreach ($entities as $entity) {
+                if ((int) \Entity::getUsedConfig('2fa_enforcement_strategy', $entity, '', true) === 1) {
```
the `true` value for default make the `get2FAEnforcement` method to return enforced=true when no enforcement has been set at all.

And so, i didn't setup the enforcement while testing the feature on an instance where the pr hasn't been tested at all, and it asked to register the 2fa application

orthagh commented on front/login.php:
```diff
@@ -86,7 +90,17 @@
 
 // now we can continue with the process...
-if ($auth->login($login, $password, (isset($_REQUEST["noAUTO"]) ? $_REQUEST["noAUTO"] : false), $remember, $login_auth)) {
+if (isset($_REQUEST['cancel'])) {
```
should we use more precise keys like `totp_cancel`, `totp_code`, `backup_code` when using POST variables between redirections ?

orthagh commented on src/Config.php:
```diff
@@ -3415,6 +3427,42 @@ public function showFormSecurity()
         echo '</td>';
         echo '</tr>';
 
+        echo '<tr class="tab_bg_1">';
```
as we target main, we may take advantage to move to twig template for this whole security form

orthagh commented on src/Entity.php:
```diff
@@ -2641,6 +2645,71 @@ public static function showUiCustomizationOptions(Entity $entity)
+
+        $rand = mt_rand();
+
+        echo "<tr><th colspan='2'>" . __('Two-factor authentication (2FA)') . "</th></tr>";
```
same, can we use twig for this part ?

orthagh commented on src/Group.php:
```diff
@@ -642,6 +647,49 @@ public function showLDAPForm($ID)
+     **/
+    public function showSecurityForm($ID)
+    {
+        $options = [];
```
same, twig ?

orthagh commented on src/Profile.php:
```diff
@@ -2150,6 +2156,60 @@ public function showFormSetupHelpdesk($openform = true, $closeform = true)
+     * @param $closeform    boolean  close the form (true by default)
+     **/
+    public function showFormSecurity($openform = true, $closeform = true)
+    {
```
same twig ?

also, the form seem (maybe there's difference) to be the same between group, profile and entity, can we factorize the display ?

orthagh commented on src/Security/PermissionManager.php:
```diff
@@ -0,0 +1,88 @@
+use ProfileRight;
+use QueryExpression;
+
+final class PermissionManager
```
can you comment above the class its general purpose, it's not that clear at read

cconard96 replied:
done. for reference, this was borrowed from the pending PR #11649

orthagh replied:
👍🏽

orthagh reviewed (changes requested):
in addition, you should add to User class the `2fa` entry to `$undisclosedFields` property. This to avoid leak of key, or backup codes

cedric-anne reviewed (changes requested):
When a user enables the 2FA on its own account, from its preferences page, backup codes are not shown after 2FA activation. They should.

cedric-anne commented:
Tests are failing and should be fixed.

PheitosaBr commented:
Dear all, congratulations on the solution. How can I install it inside my GLPI?
